File: Leetcode Daily Challenge/024-June-2025/007-lexicographically-minimum-string-after-removing-stars.cpp
// https://leetcode.com/problems/lexicographically-minimum-string-after-removing-stars/




#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    typedef pair<char, int> P;

    // TC : O(N * log(N))
    // SC : O(N)
    string clearStars(string s) {
        int n = s.length();

        auto cmp = [&](P p1, P p2) {
            if (p1.first == p2.first) {
                return p2.second > p1.second;   // higher index first
            }
            return p1.first > p2.first; // lexiographically smaller char first
        };
        priority_queue<P, vector<P>, decltype(cmp)> pq(cmp);
        
        for (int i = 0; i < n; i ++) {
            if (s[i] != '*') {
                pq.push({s[i], i});
            } else if (!pq.empty()){
                int idx = pq.top().second;
                s[idx] = '*';
                pq.pop();
            }
        }
        string result;
        for (char &ch : s) {
            if (ch != '*') {
                result += ch;
            }
        }
        return result;
    }
};

File: Leetcode Daily Challenge/024-June-2025/007-lexicographically-minimum-string-after-removing-stars.java
// https://leetcode.com/problems/lexicographically-minimum-string-after-removing-stars/




import java.util.*;

class Solution {
    // TC : O(N log N)
    // SC : O(N)
    public String clearStars(String s) {
        int n = s.length();

        char[] chars = s.toCharArray();

        // custom comparator for priority queue
        PriorityQueue<Map.Entry<Character, Integer>> pq = new PriorityQueue<>(
            (p1, p2) -> {
                if (p1.getKey().equals(p2.getKey())) {
                    return Integer.compare(p2.getValue(), p1.getValue());   // higher index first
                }
                return Character.compare(p1.getKey(), p2.getKey()); // lex char first
            }
        );

        for (int i = 0; i < n; i ++) {
            if (chars[i] != '*') {
                pq.offer(new AbstractMap.SimpleEntry<>(chars[i], i));
            } else if (!pq.isEmpty()) {
                int idx = pq.poll().getValue();
                chars[idx] = '*';
            }
        }

        StringBuilder res = new StringBuilder();
        for (char ch : chars) {
            if (ch != '*') {
                res.append(ch);
            }
        }
        return res.toString();
    }
}